src/ios/OSFirebaseCloudMessaging.swift

```diff
@@ -36,215 +32,264 @@ class OSFirebaseCloudMessaging: CDVPlugin {
-            self.sendResult(result: "", error:FirebaseMessagingErrors.obtainSilentNotificationsError as NSError, callBackID: self.callbackId)
-            return
+        guard let clearFromDatabase = command.argument(at: 0) as? Bool else {
+            return self.send(error: .obtainSilentNotificationsError, callbackId: command.callbackId)
         }
         
-        self.plugin?.getPendingNotifications(clearFromDatabase: clearFromDatabase)
+        self.commandDelegate.run { [weak self] in
+            guard let self else { return }
+            
+            do {
+                let notificationArray = try self.plugin?.getPendingNotifications()
+                let result = notificationArray?.encode()    // this is done here as notifications might be clean with the next if statement.
+                if clearFromDatabase, notificationArray?.isEmpty == false, let notificationArray {
+                    do {
+                        try self.plugin?.delete(pendingNotifications: notificationArray)
+                    } catch {
+                        throw FirebaseMessagingErrors.deleteNotificationsError
+                    }
+                }
+                self.sendSuccess(result: result, callbackId: command.callbackId)
+            } catch {
+                self.send(error: error as? FirebaseMessagingErrors ?? .obtainSilentNotificationsError, callbackId: command.callbackId)
+            }
+        }
     }
     
     @objc(unregisterDevice:)
     func unregisterDevice(command: CDVInvokedUrlCommand) {
-        self.callbackId = command.callbackId
-        Task {
-            await self.plugin?.unregisterDevice()
+        self.commandDelegate.run { [weak self] in
+            guard let self else { return }
+            
+            Task {
+                do {
+                    try await self.plugin?.deleteToken()
+                } catch {
+                    return self.send(error: .unregistrationDeleteTokenError, callbackId: command.callbackId)
+                }
+                
+                do {
+                    try await self.plugin?.unsubscribe(fromTopic: .general)
+                    self.sendSuccess(callbackId: command.callbackId)
+                } catch {
+                    self.send(error: .unregistrationError, callbackId: command.callbackId)
+                }
+            }
         }
     }
     
     @objc(getToken:)
     func getToken(command: CDVInvokedUrlCommand) {
-        self.callbackId = command.callbackId
-        Task {
-            await self.plugin?.getToken(ofType: .fcm)
+        self.commandDelegate.run { [weak self] in
+            guard let self else { return }
+            
+            Task {
+                guard let token = try? await self.plugin?.getToken()
+                else { return self.send(error: .obtainingTokenError, callbackId: command.callbackId) }
+                
+                do {
+                    try await self.plugin?.subscribe(toTopic: .general)
```

Comment:
Is this necessary? Don't we do this already in `registerDevice`?

Reply:
Ok, that makes sense. I was asking because to be honest I was confused about why this was necessary. Maybe we could come back to this topic in the future.